inlanecruising_plugin/src/inlanecruising_plugin.cpp

```diff
@@ -133,12 +133,87 @@ Eigen::Isometry2d InLaneCruisingPlugin::compute_heading_frame(const lanelet::Bas
   return carma_wm::geometry::build2dEigenTransform(p1, yaw);
 }
 
+std::pair<double, size_t> min_with_exclusions(const std::vector<double>& values, const std::unordered_set<size_t>& excluded) {
+  double min = std::numeric_limits<double>::max();
+  size_t best_idx = -1;
+  for (size_t i = 0; i < values.size(); i++) {
+    if (excluded.find(i) != excluded.end()) {
+      continue;
+    }
+
+    if (values[i] < min) {
+      min = values[i];
+      best_idx = i;
+    }
+  }
+
+  return std::make_pair(min, best_idx);
+}
+
+std::vector<double> InLaneCruisingPlugin::optimize_speed(const std::vector<double>& downtracks, const std::vector<double>& curv_speeds, double accel_limit)
+{
+  if (downtracks.size() != curv_speeds.size())
+  {
+    throw std::invalid_argument("Downtracks and speeds do not have the same size");
+  }
+
+  if (accel_limit <= 0)
+  {
+    throw std::invalid_argument("Accel limits should be positive");
+  }
+
+  bool optimize = true;
+  size_t min_index = 0;
+  std::unordered_set<size_t> visited_idx;
+  visited_idx.reserve(curv_speeds.size());
+
+  std::vector<double> output = curv_speeds;
+
+  while (optimize)
+  {
+    auto min_pair = min_with_exclusions(curv_speeds, visited_idx);
+    size_t min_idx = std::get<1>(min_pair);
+    if (min_idx == -1) {
+      break;
+    }
+
+    visited_idx.insert(min_idx); // Mark this point as visited
+
+    double v_i = std::get<0>(min_pair);
+    double x_i = downtracks[min_idx];
+    for (int i = min_idx - 1; i >= 0; i--) { // NOTE: Do not use size_t for i type here as -- with >= 0 will result in overflow
+      double v_f = curv_speeds[i];
+      double dv = v_f - v_i;
+      
+      double x_f = downtracks[i];
+      double dx = x_i - x_f; // invert dx to make math work out
+
+      if(dv > 0) {
+        v_f = std::min(v_f, sqrt(v_i * v_i + 2 * config_.max_decel * dx));
+        visited_idx.insert(i);
+      } else if (dv < 0) {
+        break;
```

Comment:
So as long as we're decelerating we don't apply the accel limit? I'm not sure if that makes sense, we'd want to smooth jerky or sudden decel as well, I'd think.

Reply:
I think at the end, we are also applying another limiter which handles it I think

Reply:
As Misheel said, the algorithm is to apply all the decelerations to min speeds first, then you apply accel limits at the end. That's to prevent your acceleration limits from overwriting your min speeds.